packages/swc/package.json

```diff
@@ -0,0 +1,73 @@
+{
+  "name": "@rollup/plugin-swc",
+  "version": "0.0.0",
+  "publishConfig": {
+    "access": "public"
+  },
+  "description": "Transpile JavaScript/TypeScript code with swc.",
+  "license": "MIT",
+  "repository": {
+    "url": "rollup/plugins",
+    "directory": "packages/swc"
+  },
+  "author": "Peter Placzek <peter.placzek1996@gmail.com>",
+  "homepage": "https://github.com/rollup/plugins/tree/master/packages/swc#readme",
+  "bugs": "https://github.com/rollup/plugins/issues",
+  "main": "dist/cjs/index.js",
+  "module": "dist/es/index.js",
+  "exports": {
+    "types": "./types/index.d.ts",
+    "import": "./dist/es/index.js",
+    "default": "./dist/cjs/index.js"
+  },
+  "engines": {
+    "node": ">=10.0.0"
+  },
+  "scripts": {
+    "build": "rollup -c",
+    "ci:coverage": "nyc pnpm test && nyc report --reporter=text-lcov > coverage.lcov",
+    "ci:lint": "pnpm build && pnpm lint",
+    "ci:lint:commits": "commitlint --from=${CIRCLE_BRANCH} --to=${CIRCLE_SHA1}",
+    "ci:test": "pnpm test -- --verbose && pnpm test:ts",
+    "prebuild": "del-cli dist",
+    "prepare": "if [ ! -d 'dist' ]; then pnpm build; fi",
+    "prerelease": "pnpm build",
+    "pretest": "pnpm build",
+    "release": "pnpm --workspace-root plugin:release --pkg $npm_package_name",
+    "test": "ava",
+    "test:ts": "tsc types/index.d.ts test/types.ts --noEmit"
+  },
+  "files": [
+    "dist",
+    "!dist/**/*.map",
+    "src",
+    "types",
+    "README.md"
+  ],
+  "keywords": [
+    "rollup",
+    "plugin",
+    "swc",
+    "speedy-web-compiler",
+    "npm",
+    "modules"
+  ],
+  "peerDependencies": {
+    "@swc/core": "^1.x",
+    "rollup": "^2.x || ^3.x"
+  },
+  "peerDependenciesMeta": {
+    "rollup": {
+      "optional": true
+    }
+  },
+  "dependencies": {
+    "smob": "^0.0.6"
+  },
+  "devDependencies": {
+    "@swc/core": "^1.3.27",
```

Comment:
I wonder if this should be a dependency instead of bundling it in? 🤔

Reply:
I think we should avoid bundling external libraries, unless it is an end application. This prevents that a library has already been bundled several times via different third-party libraries and is therefore redundant in the final bundle.

Reply:
Yep in which case this should be in `dependencies`?

Reply:
It is defined as `devDependency` and `peerDependency` so it is exluded for bundling. This is a trick to avoid linting and compiler issues, when importing different components and you are still able to let the end user have a arbitary `v1.3.x` version of the `@swc/core` package.

Reply:
Ah I missed that it was a peer dependency

Personally I’m not a fan of peer dependencies as it puts the burden on the user to install it and make sure it stays within the required version range. When it isn’t a valid version it usually ends up as a warning amongst a load of other warnings that get missed

Also given the plugin requires the dependency in order to function is it really a peer dependency?

As a standard dependency there is still no duplication, unless the user also needs a different version for something else (which would work fine). Also on install it would normally resolve to the latest version and the user can update it when they update their lock file.

Maybe let’s see what others think.

If it does stay as a peer dependency I think we need to document it in the readme and it would be good to throw an error asking the user to add the peer dependency if it is missing

Reply:
> Ah I missed that it was a peer dependency
> 
> Personally I’m not a fan of peer dependencies as it puts the burden on the user to install it (I think depending on the package manager?) and make sure it stays within the required version range. When it isn’t a valid version it usually ends up as a warning amongst a load of other warnings that get missed
> 
> Also given the plugin requires the dependency in order to function is it really a peer dependency?
> 
> As a standard dependency there is still no duplication, unless the user also needs a different version for something else (which would work fine). Also on install it would normally resolve to the latest version and the user can update it when they update their lock file.
> 
> Maybe let’s see what others think.
> 
> If it does stay as a peer dependency I think we need to document it in the readme and it would be good to throw an error asking the user to add the peer dependency if it is missing

Peer dependencies are automatically installed by the package manager and has many advantages. The problem with dependencies is that from v1+, only patches within the minor range are allowed. So we would always have to manually patch the version when security updates or new features are implemented. 
Only peer dependencies which are declared as **required: 'optional' ** in the `peerDependenciesMeta` section of the package.json file are not installed. 
I think it would be good if you would have another look at the npm and especially package.json features, because there are already cool gadges.

Reply:
Cool I will have a read about `peerDependenciesMeta` when I get chance :)

What about npm in node lts and other popular package managers? Do they follow the same spec and auto install?

Not following the version thing. If it’s a peer dependency and the user is using a lock file they have to choose when to update it, and if it’s a dependency, they also choose when to update it by running `npm update`. That would update the nested dependencies right (or maybe I remembered wrong)

Reply:
I know that yarn supports it too. They also give a neat example https://classic.yarnpkg.com/lang/en/docs/dependency-types/#toc-peerdependencies
You got a good poin there with `npm update` :blush: , but this only applies for "end" libraries. Lock files are not passed when publishing, so everyone in the chain has to do the update process again, but otherwise you are of course right.

Reply:
Yeh so if iiuc dependency or peer dependency shouldn’t effect how frequently this plugin needs to be updated. With either option you can still configure a range with semver

Would be good to know what node lts does out of the box and if it has the same behaviour as latest

Reply:
But I would like to focus more on the actual content, if that's okay with you, and not on whether a particular dependency should be managed as a dependency or peer-dependency, which is basically a general design question and has nothing to do with the plugin request itself.

Reply:
> Yeh so if iiuc dependency or peer dependency shouldn’t effect how frequently this plugin needs to be updated. With either option you can still configure a range with semver
> 
> Would be good to know what node lts does out of the box and if it has the same behaviour as latest

Node 16 (npm 7) definitly supports it, and regarding the official node repo, Node 16 is already not longer in LTS mode. https://github.com/nodejs/Release

Reply:
> But I would like to focus more on the actual content, if that's okay with you, and not on whether a particular dependency should be managed as a dependency or peer-dependency, which is basically a general design question and has nothing to do with the plugin request itself

I think it’s important to get this right as changing it later is possible but probably a breaking change

I’m happy to go with peer dependency if others also think this makes sense. I may also agree with you but want to do a bit of reading first on how peer dependencies work now first

Given they definitely didn’t use to install automatically and there may be people on older versions I think throwing a helpful error if the package is missing is probably worthwhile

Reply:
So essentially I don’t think this thread should be resolved unless another maintainer understands this and agrees it makes sense (which may be me once I’ve had chance to read up on it more) :)

Reply:
Ok done a bit more googling

>I know that yarn supports it too. They also give a neat example https://classic.yarnpkg.com/lang/en/docs/dependency-types/#toc-peerdependencies

>Having a peer dependency means that your package needs a dependency that is the same exact dependency as the person installing your package. This is useful for packages like react that need to have a single copy of react-dom that is also used by the person installing it.

I don't think that really applies in this use case. We don't _need_ the user to have the same version of `@swc/core` or even use it themselves. If they had to construct an instance of `@swc/core` and pass it in then I think that could apply.

The [docs for `peerDepdenciesMeta`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#peerdependenciesmeta) say

>When a user installs your package, npm will emit warnings if packages specified in peerDependencies are not already installed. The peerDependenciesMeta field serves to provide npm more information on how your peer dependencies are to be used. Specifically, it allows peer dependencies to be marked as optional.

>Marking a peer dependency as optional ensures npm will not emit a warning if the soy-milk package is not installed on the host. This allows you to integrate and interact with a variety of host packages without requiring all of them to be installed.

Which reads to me like npm auto installing a peer dependency is not the ideal outcome. If it installs, but still emits a warning for the user, that's not ideal.

And docs for [`peerDependencies`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#peerdependencies)

>In some cases, you want to express the compatibility of your package with a host tool or library, while not necessarily doing a require of this host. This is usually referred to as a plugin. Notably, your module may be exposing a specific interface, expected and specified by the host documentation.

We _are_ doing an import of `@swc/core` ourselves. I don't think `@swc/core` is a plugin for the rollup swc package.

>Node 16 (npm 7) definitly supports it, and regarding the official node repo, Node 16 is already not longer in LTS mode. https://github.com/nodejs/Release so we are fine ✌️

Thanks for checking. At least supported npm versions would be consistent if we did this.

I still don't completely understand what you meant by

>The problem with dependencies is that from v1+, only patches within the minor range are allowed.

? If we use `^1.3.27` that would mean users can get 1.3.27 and any v1 version later than that, without us having to publish an update to this package.

So my opinion is this should be a `dependency` not `peerDependency`

Reply:
> Ok done a bit more googling
> 
> > I know that yarn supports it too. They also give a neat example https://classic.yarnpkg.com/lang/en/docs/dependency-types/#toc-peerdependencies
> 
> > Having a peer dependency means that your package needs a dependency that is the same exact dependency as the person installing your package. This is useful for packages like react that need to have a single copy of react-dom that is also used by the person installing it.

I think you have misunderstood. Maybe check out the following articles/sections.
https://nodejs.org/es/blog/npm/peer-dependencies/
https://docs.npmjs.com/cli/v9/configuring-npm/package-json#peerdependencies

> I don't think that really applies in this use case. We don't _need_ the user to have the same version of `@swc/core` or even use it themselves. If they had to construct an instance of `@swc/core` and pass it in then I think that could apply.
> 
> The [docs for `peerDepdenciesMeta`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#peerdependenciesmeta) say
> 
> > When a user installs your package, npm will emit warnings if packages specified in peerDependencies are not already installed. The peerDependenciesMeta field serves to provide npm more information on how your peer dependencies are to be used. Specifically, it allows peer dependencies to be marked as optional.
> 
> > Marking a peer dependency as optional ensures npm will not emit a warning if the soy-milk package is not installed on the host. This allows you to integrate and interact with a variety of host packages without requiring all of them to be installed.
> 
> Which reads to me like npm auto installing a peer dependency is not the ideal outcome. If it installs, but still emits a warning for the user, that's not ideal.

For npm versions less than v7 a warning is issued and for versions greater than v7 the peer dependency is automatically installed if not already present.
https://github.blog/2021-02-02-npm-7-is-now-generally-available/#peer-dependencies

> And docs for [`peerDependencies`](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#peerdependencies)
> 
> > In some cases, you want to express the compatibility of your package with a host tool or library, while not necessarily doing a require of this host. This is usually referred to as a plugin. Notably, your module may be exposing a specific interface, expected and specified by the host documentation.
> 
> We _are_ doing an import of `@swc/core` ourselves. I don't think `@swc/core` is a plugin for the rollup swc package.

We are also doing an import on `rollup` in nearly every package of this monorepo and rollup is listed in every package.json as peer-dependency. Soa this is not an argument.

> > Node 16 (npm 7) definitly supports it, and regarding the official node repo, Node 16 is already not longer in LTS mode. https://github.com/nodejs/Release so we are fine ✌️
> 
> Thanks for checking. At least supported npm versions would be consistent if we did this.
> 
> I still don't completely understand what you meant by
> 
> > The problem with dependencies is that from v1+, only patches within the minor range are allowed.
> 
> ? If we use `^1.3.27` that would mean users can get 1.3.27 and any v1 version later than that, without us having to publish an update to this package.
> 
> So my opinion is this should be a `dependency` not `peerDependency`

Reply:
By the way, the same is true for the babel plugin. So I really don't understand why you criticize the plugin but not there. This makes no sense to me. 
https://github.com/rollup/plugins/blob/master/packages/babel/package.json
Since this procedure is also the case with other transpiler plugins such as babel, I will not go into further detail here.

Reply:
>We are also doing an import on rollup in nearly every package of this monorepo and rollup is listed in every package.json as peer-dependency. Soa this is not an argument.

In this package the rollup import is just for types, which is just for dev, which is why it's a dev dependency. It also being a peer dependency does makes sense to me, because this plugin is adhering to the rollup plugin api, which may change in a future rollup major version

>By the way, the same is true for the babel plugin. So I really don't understand why you criticize the plugin but not there. This makes no sense to me.
https://github.com/rollup/plugins/blob/master/packages/babel/package.json

Good point. If that plugin was being added now I'd be making the same points, but that ship has sailed :)

My view after reading the docs is that this should be a `dependency`, for the reasons above.

As I've said before If other maintainers think it's fine as a peer dependency, sure lets go with that. @rollup/plugin-maintainers thoughts?

Please stop resolving this thread though. IMO it's not resolved yet.

Reply:
Nevertheless, I would like to thank you for the time you have invested in the investigation and I am also well aware that you have the best intentions :relaxed:  . Thank you very much for that :v: . I only temporarily had the feeling that we were going around in circles when discussing peer dependency or rather dependency :see_no_evil: . But I also appreciate it very much to exchange ideas about such things. Maybe we get to work together on one or the other project / package independently of the review process :sunglasses: .

Reply:
Stepping in to be the tie-breaker, per request:

> By the way, the same is true for the babel plugin. So I really don't understand why you criticize the plugin but not there. This makes no sense to me.

It is never invalid to point out current issues even if those issues previously exist.

You're both making valid points, but there is needed nuance. Questions that lend to general guidance on this kind of thing:

- is the dependency _HUGE_?
- is the dependency pure JavaScript?
- is the dependency platform-depdendent?

### is the dependency _HUGE_?

This is a no-brainer. If the dependency is _huge_ like Babel or TypeScript, then we set it as a peer and let the user install the correct version within range. We don't want to be in the business of bundling typescript with the typescript plugin only to have the user also have typescript in their repo and bloat it further.

### is the dependency pure JavaScript?

If yes, then it's probably safe to bundle and it's one less thing the user has to worry about, but we also want to make sure we document what version of the target dependency that the plugin currently supports. `sucrase` and `buble` do this.

### is the dependency platform-depdendent?

If any of our dependencies need to be built for the platform of the machine the user is running on, we need to leave them as peers.

And this is where both of your arguments fall short. Within the `@swc/core` package there is a post-install script which runs and compiles bindings _which is platform dependent_. That means that there is no way we can reasonably bundle `swc` in this plugin so that it works on all platforms. We might get lucky and it just works, but with any dependency which has node bindings (or binaries) compiled as part of an install step, we need to leave that as something the user is responsible for.

WRT peers being auto-installed or not, documentation fixes this. If we list what the requirements are in documentation, then the user has recourse if they are not auto-installing peers. PNPM, NPM, and Yarn all do this now and all have settings to prevent this. So, should a user disable auto-installing peers (as I do) they need to look to the docs. We have a `Requirements` section in each README right at the top for exactly this purpose.

So, in summary, for the reasons listed above, `swc` dependencies should be `peerDependencies` for this plugin.

Reply:
Different package managers (npm, pnpm, yarn, etc.) and different versions of those package managers all handle `peerDependencies` differently. Very often the user must add the dependency to their `package.json` and install it though it sounds like the latest version of npm will do it for you.

I'm not quite following the argument that native libraries must be `peerDependencies`. I've not seen the installation of swc fail before and I'm not sure if installing swc fails how that would differ if it's in `dependencies` vs `peerDependencies` and whether they would need to follow the same steps or different steps in each case.

If it's helpful to see what others have done, Vite added it in `dependencies`: https://github.com/vitejs/vite-plugin-react-swc/blob/0d6588e76471bf5a801a0cab14808dafc78e7d09/package.json#L19

And it's done the same with `esbuild` as well: https://github.com/vitejs/vite/blob/356ddfe2e534800c1fd59fa502a2c4c8945f4f92/packages/vite/package.json#L69

I've not heard any complaints about this or instances of these libraries failing though there are probably others who would be more tuned in.

To me you use `peerDependencies` when you need the user to control the exact version being used or want them to provide the library. E.g. plugins declaring a dependency on the main package are a great use case for `peerDependencies`. I'd find making it a peer dependency in this case to be annoying because it's a package I wouldn't already have installed and would double the setup steps.

Personally, I would say to put it in `dependencies` though I don't really want to create a tie again and @shellscape's vote counts more than mine.

Reply:
The kicker is that we're attempting to bundle it. That's the differentiator. If we weren't bundling this plugin to ship it to NPM, `dependencies` would be correct.

Reply:
I agree I would definitely not bundle it. Can we exclude that dependency from bundling and list it in `dependencies`?

Reply:
Sorry I haven’t caught up, but to be clear I was proposing `dependencies` over `peerDependencies`.

I don’t think we should bundle it in and just put it in the dev list

Reply:
Excellent question. Probably? https://www.rollupjs.org/configuration-options/#external

Reply:
I think if we just put it in `dependencies` the `createConfig` helper would already configure `external` to include it

Reply:
yep. that's correct https://github.com/rollup/plugins/blob/69146cdf1f7353aa72364f69fa5e49c732a1abd8/shared/rollup.config.mjs#L15

That goes back to #363 which I didn't have a hand in. But there we go. It goes into `dependencies` and we're automatically not bundling it.

Reply:
Since consistency is one of my basic principles, I would like us to take the same approach for `babel`. In general, the reasoning why it is dependency instead of peer-dependency has not convinced me much. But since @shellscape is code owner, I will follow his decision. 
To avoid arbitrariness or making someone additional work, I created a pull request for babel:
https://github.com/rollup/plugins/pull/1423